pkg/apis/autoscaling/types.go

```diff
@@ -95,6 +95,68 @@ type HorizontalPodAutoscalerSpec struct {
 	// more information about how each type of metric must respond.
 	// +optional
 	Metrics []MetricSpec
+
+	// Behavior contains settings to specify the scaling behavior both while scaling up
+	// and scaling down. It also is used to customize other aspects of scaling like the
+	// scale down stabilization window
+	// +option
+	Behavior *HorizontalPodAutoscalerBehavior
+}
+
+// HorizontalPodAutoscalerBehavior configures a scaling behavior for Up and Down direction
+// (scaleUp and scaleDown fields respectively)
+type HorizontalPodAutoscalerBehavior struct {
+	// The scaling behavior while scaling up. Contains a list of policies and the policy selector.
+	// +optional
+	ScaleUp *HPAScalingRules
+	// The scaling behavior while scaling down. Contains a list of policies and the policy selector.
+	// +optional
+	ScaleDown *HPAScalingRules
+}
+
+// ScalingPolicySelect is used to specify which policy should be used while scaling in a certain direction
+type ScalingPolicySelect string
+
+const (
+	// MaxPolicySelect selects the policy with the highest possible change
+	MaxPolicySelect ScalingPolicySelect = "max"
+	// MinPolicySelect selects the policy with the lowest possible change
+	MinPolicySelect ScalingPolicySelect = "min"
+)
+
+// HPAScalingRules configures the scaling policy and the policy selector
+type HPAScalingRules struct {
+	// StabilizationWindowSeconds is the number of seconds for which past recommendations should be considered while
+	// scaling down. This prevents scale down of the autoscaling target in case the load fluctuates.
+	// +optional
+	StabilizationWindowSeconds *int32
+	// SelectPolicy is used to specify which policy should be selected after evaluation.
+	// +optional
+	SelectPolicy *ScalingPolicySelect
+	// Policies is a list of potential scaling polices which can used during scaling.
+	// +optional
+	Policies []HPAScalingPolicy
+}
+
+// HPAScalingPolicyType is the type of the policy which could be used while making scaling decisions
+type HPAScalingPolicyType string
+
+const (
+	// PodsScalingPolicy is a policy used to specify a change in absolute number of pods.
+	PodsScalingPolicy HPAScalingPolicyType = "pods"
+	// PercentScalingPolicy is a policy used to specify a relative amount of change with respect to
+	// the current number of pods.
+	PercentScalingPolicy HPAScalingPolicyType = "percent"
+)
+
+// HPAScalingPolicy is a single policy which must hold true for a specified past interval.
+type HPAScalingPolicy struct {
+	// Type is used to specify the scaling policy.
+	Type HPAScalingPolicyType
+	// Value contains the amount of change which is permitted by the policy.
+	Value *int32
+	// PeriodSeconds specifies the window of time for which the policy should hold true.
+	PeriodSeconds *int32
```

Comment:
I would prefer to make these `int32` (no pointer) as in the KEP.  This indicates they are required.  An incomplete `HPAScalingPolicy` should be invalid (defaults at this level would be confusing).

![image](https://user-images.githubusercontent.com/1103629/66809105-fd5f6e00-ef2c-11e9-9003-8aa2f75f843a.png)

Reply:
@gliush @arjunrn and I discussed this in Slack.  We concluded that `Value` and `PeriodSeconds` should indeed be `int32` (no pointer).  Our intent is that a user must provide all parts of the policy explicitly (no defaults) to make the API less confusing.  So we will validate that `Value` and `PeriodSeconds` are **not** `0`.  And that `Type` is not `""`.

The open question was whether a `Value` of `0` is a valid use case.  It's included in the KEP as one: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md#story-4-scale-up-as-usual-do-not-scale-down.  We think it could be, so the API should allow for it.

However we prefer not to use pointers.  They are cumbersome in golang because they require an intermediate variable.  They add a layer of indirection and risk for nil dereference.  And they make the API less intuitive (e.g. what does `nil` mean?).  We decided that a `SelectPolicy` value of `"disabled"` would be a more intuitive way of turning off up or down scaling.  And would allow us to avoid a `*int32` for `Value`.

For `PeriodSeconds`, `0` simply doesn't make sense (divide by 0).  So it doesn't need a pointer either.

Reply:
Just one note here. For "Type" we check that it has an allowed value ("max", "min" or "disabled"), and not a simple `is not ""`